### ISSUE TITLE
Is there an option to configure reconnect attempt limit?

### ISSUE THREAD

**Issue by manojp1988** [stomp-js/ng2-stompjs#77]
Hi, In my use case, I am planning to attempt reconnecting for 5 times, after then, I'm planning to redirect the user to login page.

Currently, I am stompService.state and listen for CLOSED status and I'm redirecting to login page. But if I see the logs, connection attempt is still progressing. 

I need some way to stop attempting to reconnect.

**Comment by kum-deepak** [comment]
Ideally calling `client.disconnect` should have stopped the reconnect process. However there is a bug in the underlying library which will likely break part of it.

Give me couple of days, I will attempt to fix underlying bug. After that calling `client.disconnect` will stop re-connection attempts.

**Comment by kum-deepak** [comment]
I have upgraded the underlying library `@stomp/stompjs` to `4.0.7`. Please force a dependency to that version in you `package.json`, I will suggest:

```JSON
{
  "dependencies": {
    "@stomp/stompjs": "^4.0.0 >=4.0.7"
  }
}
``` 

Once you have done that calling `client.disconnect()` should stop re-connection attempts. Let me know if it resolves the issue.

**Comment by kum-deepak** [comment]
See also #78, fundamentally the same underlying cause.

**Comment by manojp1988** [comment]
No, below is my code. It is still attempting to reconnect.  ShowAlert will show a popup on clicking 'Ok', the callback method will be called.

```ts
 this._stompService.state
                    .pipe(
                        map((state: number) => StompState[state]),
                        filter(status => status === 'CLOSED'),
                        take(RECONNECT_LIMIT)
                    )
                    .subscribe((status: string) => {
                        this.attemptLimit--;
                        if (this.attemptLimit > 0) {
                            return;
                        }
                        this.store.dispatch(
                            new fromRootActions.ShowAlert({
                                message: 'Socket connection failure.',
                                error: 'Connection Error',
                                callback: () => {
                                    this.stompSubscription.forEach(subscription => subscription.unsubscribe());
        							this._stompService.disconnect();
                                    this.attemptLimit = RECONNECT_LIMIT;
                                    this.store.dispatch(new fromAuthActions.Logout());
                                }
                            })
                        );
                    });
`

**Comment by manojp1988** [comment]
by the way, if you add a config to set reconnect limit, it would be good.

**Comment by kum-deepak** [comment]
This code will not work. I will send some sample code that is likely to work. 
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by manojp1988** [comment]
Thank you. If you point what's wrong, that would be more helpful

**Comment by kum-deepak** [comment]
On every disconnect you have to reduce your count, once reached 0, you will need to call disconnect.

On successful connect the count needs to be reset.

Just doing take(N), will not stop reconnect attempts. 
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by manojp1988** [comment]
I just added take(N) for observable auto unsubscription. In my case, this._stompService.disconnect(); is called but not sure why it is still trying to reconnect.

**Comment by kum-deepak** [comment]
I have added a new branch to one of the samples https://github.com/stomp-js/ng6-stompjs-demo/tree/reconnect-retry.

You can easily view the changes as the pull request https://github.com/stomp-js/ng6-stompjs-demo/pull/1

In this process also identified a bug in disconnect code in StompService which is locally fixed in this sample: https://github.com/stomp-js/ng6-stompjs-demo/pull/1/commits/b8436ce306528364667f6c7c6bb52d56c6f56b28

The actual sample code can be seen at https://github.com/stomp-js/ng6-stompjs-demo/pull/1/commits/18855b78584955b3e947157e4d63e462c98935d0

Please consider this as a sample as it does not cover all possible life cycle properly.

I am currently in the process of rewriting the underlying library to make this and similar features easier to implement. You can contribute/track progress at https://github.com/stomp-js/stompjs.

**Comment by manojp1988** [comment]
I'm calling stompRService.disconnect() once retry limit reached.  The disconnect method will actually run if client is already connected. If not, it won't disconnect. And the example attached was working because, this disconnect method is overridden.

I feel this should be handled in the library side instead of overriding the disconnect method in user's code.

```ts
  /**
   * It will disconnect from the STOMP broker.
   */
  public disconnect(): void {

    // Disconnect if connected. Callback will set CLOSED state
    if (this.client && this.client.connected) {
      // Notify observers that we are disconnecting!
      this.state.next(StompState.DISCONNECTING);

      this.client.disconnect(
        () => this.state.next(StompState.CLOSED)
      );
    }
  }
```

**Comment by kum-deepak** [comment]
Please check if the code works, based on your confirmation, I will apply this change onto the library and make a new release 😄

**Comment by manojp1988** [comment]
Hi, is that disconnect() method change i suggested released?

**Comment by kum-deepak** [comment]
Released 6.0.1 (for Angular6) and 4.0.1 (For Angular4/5).

Please check and confirm back.

**Comment by Halynsky** [comment]
I have the same issue with this versions:
```
    "@stomp/ng2-stompjs": "6.0.1",
    "@stomp/stompjs": "4.0.8",
```
Disconect method of StompRService doesn't stop recconcetions.

**Comment by kum-deepak** [comment]
Please save your console output as a file and attach. Thanks!

**Comment by Halynsky** [comment]
```
  this.securityService.onLogout.subscribe(() => {
      this.stompRService.disconnect();
    });
```
After User logout stomp still trying to reconnect.
Browser log is attached but I don't think that it will help you.
[localhost-1538059008566.log](https://github.com/stomp-js/ng2-stompjs/files/2424642/localhost-1538059008566.log)

And I have a question. Why you don't update @stomp/stompjs to the latest version in the @stomp/ng2-stompjs dependancies.

**Comment by kum-deepak** [comment]
Thanks your logfile is helpful.

`@stomp/stompjs` new version is a bottom up rewrite - it is in beta. One of the key motivation for that rewrite has been this issue.

There is compatibility mode in `@stomp/stompjs`, so, however the TypeScript type lib has non compatible changes. So, it is not a drop-in replacement.

I am planning to release a beta of this library pointing to beta of 5.0.0 of  `@stomp/stompjs`.

BTW, new version of `@stomp/stompjs` has a `beforeConnect` callback to handle the updating token use case.

I will open a new issue to track progress of the upgrade

**Comment by kum-deepak** [comment]
I have released 7.0.0 of this library. It (and the underlying @stomp/stompjs 5.0.0) has a beforeConnect method, you can change configuration or even call `deactivate` (newer name for `disconnect`).

I will write some sample code in coming weeks to demonstrate these.
